src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java

```diff
@@ -0,0 +1,241 @@
+
+			if (info.firstType.isAssignableFrom(firstType)) { // Attempt to convert the second argument to the second comparator type
+				Converter<Type2, CType2> sc2 = Converters.getConverter(secondType, info.secondType);
+				if (sc2 != null)
+					return new ConvertedComparator<>(null, info.comparator, sc2);
+			}
+
+			if (info.secondType.isAssignableFrom(secondType)) { // Attempt to convert the first argument to the first comparator type
+				Converter<Type1, CType1> fc1 = Converters.getConverter(firstType, info.firstType);
+				if (fc1 != null)
+					return new ConvertedComparator<>(fc1, info.comparator, null);
+			}
+
+		}
+
+		// Attempt converting one parameter but with reversed types
+		for (ComparatorInfo<?, ?> unknownInfo : COMPARATORS) {
+			if (!unknownInfo.comparator.supportsInversion()) // Unsupported for reversing types
+				continue;
+
+			ComparatorInfo<CType1, CType2> info = (ComparatorInfo<CType1, CType2>) unknownInfo;
+
+			if (info.secondType.isAssignableFrom(firstType)) { // Attempt to convert the second argument to the first comparator type
+				Converter<Type2, CType1> sc1 = Converters.getConverter(secondType, info.firstType);
+				if (sc1 != null)
+					return new InverseComparator<>(new ConvertedComparator<>(sc1, info.comparator, null));
+			}
+
+			if (info.firstType.isAssignableFrom(secondType)) { // Attempt to convert the first argument to the second comparator type
+				Converter<Type1, CType2> fc2 = Converters.getConverter(firstType, info.secondType);
+				if (fc2 != null)
+					new InverseComparator<>(new ConvertedComparator<>(null, info.comparator, fc2));
+			}
+
+		}
+
+		// Attempt converting both parameters
+		for (ComparatorInfo<?, ?> unknownInfo : COMPARATORS) {
+			ComparatorInfo<CType1, CType2> info = (ComparatorInfo<CType1, CType2>) unknownInfo;
+
+			Converter<Type1, CType1> c1 = Converters.getConverter(firstType, info.firstType);
+			Converter<Type2, CType2> c2 = Converters.getConverter(secondType, info.secondType);
+			if (c1 != null && c2 != null)
+				return new ConvertedComparator<>(c1, info.comparator, c2);
+
+		}
+
+		// Attempt converting both parameters but with reversed types
+		for (ComparatorInfo<?, ?> unknownInfo : COMPARATORS) {
+			if (!unknownInfo.comparator.supportsInversion()) // Unsupported for reversing types
+				continue;
+
+			ComparatorInfo<CType1, CType2> info = (ComparatorInfo<CType1, CType2>) unknownInfo;
+
+			Converter<Type1, CType2> c1 = Converters.getConverter(firstType, info.secondType);
+			Converter<Type2, CType1> c2 = Converters.getConverter(secondType, info.firstType);
+			if (c1 != null && c2 != null)
+				return new InverseComparator<>(new ConvertedComparator<>(c2, info.comparator, c1));
+
+		}
+
+		// Same class but no comparator
+		if (firstType != Object.class && secondType == firstType) {
+			return (Comparator<Type1, Type2>) EQUALS_COMPARATOR;
+		}
+
+		// Well, we tried!
+		return null;
```

Comment:
You could probably implement a 'default' comparator here (for the purposes of equals/sorting checks) that simply compares by equals and hashCode.

Reply:
I don't think that's a good idea, Skript currently has an error if two objects cannot be compared, e.g. a number and a string. Returning a default comparator will break this